content/en/blog/_posts/2024-11-11-dynamic-kubernetes-api-server-for-cozystack/index.md

```diff
@@ -0,0 +1,454 @@
+controllers that handle your business logic for reconciling and managing these kinds of resources.
+This approach is well-documented, with a wealth of information available online on how to develop your
+own operators.
+
+However, this is [not the only way](https://kubernetes.io/docs/concepts/extend-kubernetes/#api-extensions)
+to extend the Kubernetes API. For more complex scenarios such as implementing imperative logic,
+managing subresources, and dynamically generating responses—the Kubernetes API Aggregation Layer
+provides an effective alternative. Through the Aggregation Layer, you can develop a custom
+extension api-server and seamlessly integrate it within the broader Kubernetes API framework.
+
+In this article, we will explore the API Aggregation Layer, the types of challenges it is well-suited
+to address, cases where it may be less appropriate, and how we utilized this model to implement
+our own extension API server in Cozystack.
+
+## What Is the API Aggregation Layer?
+
+First, let's get definitions straight to avoid any confusion down the road.
+The [API Aggregation Layer](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/apiserver-aggregation/)
+is a feature in Kubernetes, while an extension api-server is a specific implementation of an
+api-server for the Aggregation Layer. An extension api-server is just like the standard Kubernetes api-server
+, except it runs separately and handles requests for your specific resource types.
+
+So, the Aggregation Layer lets you write your own extension api-server, integrate it easily into Kubernetes,
+and directly process requests for resources in a certain group. Unlike the CRD mechanism, the api-server
+is registered in Kubernetes as an `APIService`, telling Kubernetes to consider this new api-server and acknowledge
+that it serves certain API resources. Here's a simple example:
+
+```shell
+# kubectl get apiservices.apiregistration.k8s.io
+NAME                          	SERVICE                   	AVAILABLE   AGE
+v1alpha1.apps.cozystack.io    	cozy-system/cozystack-api 	True    	7h29m
+```
+
+As soon as the Kubernetes api-server receives requests for resources in the group
+`v1alpha1.apps.cozystack.io`, it redirects all those requests to our extension api-server,
+which can handle them based on the business logic we've built into it.
+
+
+## When to use the API Aggregation Layer
+
+The API Aggregation Layer helps solve several issues where the usual CRD mechanism might
+not enough. Let's break them down.
+
+### Imperative Logic and Subresources
+
+Besides regular resources, Kubernetes also has something called subresources.
+
+{{% alert title="Note" color="info" %}}
+Subresources are additional actions or operations you can perform on primary resources
+(like Pods, Deployments, Services) via the Kubernetes API. They provide interfaces to manage
+specific aspects of resources without affecting the entire object.
+{{% /alert %}}
```

Comment:
```suggestion
In Kubernetes, subresources are additional actions or operations you can perform on primary resources
(like Pods, Deployments, Services) via the Kubernetes API. They provide interfaces to manage
specific aspects of resources without affecting the entire object.
```
I don't really like this though. We prefer subresources to be more noun like. For example, the `status` subresource is not really an action.